baseplate/queue_consumer.py

```diff
@@ -0,0 +1,127 @@
+
+    The ``handler`` function must take 3 arguments:
+
+    * ``context``: a baseplate context
+    * ``message_body``: the text body of the message
+    * ``message``: :py:class:`kombu.message.Message`
+
+    The consumer will automatically ack each message after the handler method
+    exits. If there is an error in processing and the message must be retried
+    the handler should raise an exception to crash the process. This will
+    prevent the ack and the message will be requeued at the head of the queue.
+
+    :param baseplate.core.Baseplate baseplate: A baseplate instance for the
+        service.
+    :param kombu.Exchange exchange:
+    :param kombu.connection.Connection connection:
+    :param str queue_name: The name of the queue.
+    :param list routing_keys: List of routing keys.
+    :param handler: The handler method.
+
+    """
+    queues = []
+    for routing_key in routing_keys:
+        queue = Queue(
+            name=queue_name,
+            exchange=exchange,
+            routing_key=routing_key,
+        )
+        queues.append(queue)
+
+    print("registering %s as a handler for %r" % (handler.__name__, queues))
+    kombu_consumer = KombuConsumer(baseplate, connection, exchange, queues)
+    kombu_consumer.start()
+
+    print("waiting for messages")
+    while True:
+        with kombu_consumer.get_message() as (context, message, span):
+            handler(context, message.body, message)
+            message.ack()
+
+
+class _ConsumerWorker(ConsumerMixin):
+    def __init__(self, connection, queues, work_queue):
+        self.connection = connection
+        self.queues = queues
+        self.work_queue = work_queue
+
+    def get_consumers(self, Consumer, channel):
+        return [Consumer(
+            queues=self.queues,
+            on_message=self.on_message,
+        )]
+
+    def on_message(self, message):
+        self.work_queue.put(message)
+
+
+class ConsumerContext(object):
+    pass
+
+
+class KombuConsumer(object):
+    def __init__(self, baseplate, connection, exchange, queues):
+        self.baseplate = baseplate
+        self.connection = connection
+        self.exchange = exchange
+        self.queues = queues
+        self.work_queue = queue.Queue()
+        self.consumer = _ConsumerWorker(connection, queues, self.work_queue)
+        self.thread = None
+
+    def start(self):
+        assert self.thread is None, "consumer already started"
+        self.thread = Thread(target=self.consumer.run)
+        self.thread.name = "consumer message pump"
+        self.thread.daemon = True
+        self.thread.start()
+
+    @contextlib.contextmanager
+    def get_message(self, timeout=None):
+        with self.get_batch(max_items=1, timeout=timeout) as (context, batch, span):
+            yield context, batch[0], span
+
+    @contextlib.contextmanager
+    def get_batch(self, max_items, timeout):
+        assert self.thread is not None, "you must start the consumer first"
```

Comment:
Rather than requiring this object be in a specific state for some of its methods to be used, how about having a consumer factory that does what `start()` currently does and returns an object which has `get_message` / `get_batch` on it? That way the user is funnelled to doing it the right way by the types involved.

Reply:
I tried doing something like this:

```
kombu_consumer_factory = KombuConsumerFactory(connection, exchange, queues)
baseplate.add_to_context(queue_name, kombu_consumer_factory)

while True:
    context = ConsumerContext()
    with baseplate.make_server_span(context, queue_name) as span:
        queue = getattr(context, queue_name)
        message = queue.get_message()
        handler(context, message.body, message)
        message.ack()
```

but this doesn't quite work because when `baseplate.make_server_span` is called it creates a KombuConsumer instance with a work_queue, and all the messages are put in that queue, but the next time through the loop we call `baseplate.make_server_span` which makes another instance of KombuConsumer with a different work_queue, but all the messages are in the original work_queue.

I can flip it around like:

```
kombu_consumer_factory = KombuConsumerFactory(connection, exchange, queues)
baseplate.add_to_context(queue_name, kombu_consumer_factory)

context = ConsumerContext()
with baseplate.make_server_span(context, queue_name) as span:
    queue = getattr(context, queue_name)
    while True:
        message = queue.get_message()
        handler(context, message.body, message)
        message.ack()
```

which works, but now we're using the same span to process many messages, which seems wrong (in zipkin this will look like a giant span that includes the processing of hundreds or thousands of messages).

What do you think?

Reply:
Would it be better to make `start` private and just call `_start` on the first call to `get_batch`?

Alternatively, if we prefer the factory route, I don't think we want to use a `ContextFactory`, just a regular `Factory` that `starts` the consumer before returning it, I think the idea is to break KombuConsumer into 2 classes like:

```python
class KombuConsumerFactory(object):
    def __init__(self, baseplate, connection, exchange, queues):
        ...

    def build_consumer(self):
        work_queue = queue.Queue()
        consumer = _ConsumerWorker(self.connection, self.queues, self.work_queue)
        thread = Thread(target=consumer.run)
        thread.name = "consumer message pump"
        thread.daemon = True
        thread.start()
        return KombuConsumer(self.baseplate, self.queues, work_queue, thread)

class KombuConsumer(object):
    def __init__(self, baseplate, queues, work_queue, thread):
        ...

    @contextlib.contextmanager
    def get_message(self, timeout=None):
        ...

    @contextlib.contextmanager
    def get_batch(self, max_items, timeout):
        ...
```

Reply:
Ah yeah, sorry for the Factory/ContextFactory confusion. Just a normal factory is all I meant!

Splitting it up like that just means you can't misuse it and it makes things a bit more testable since the worker thread can be mocked out.